### PR TITLE
implement brotli classes

### PR DESCRIPTION
**Pull request by anonrig**
Implements:
- `new BrotliCompress()`
- `new BrotliDecompress()`
- `createBrotliDecompress()`
- `createBrotliCompress()`
functions and classes to `node:zlib`

### CODE REVIEW

**Comment by npaun** [comment]
GH won't let me comment on the line, but turn on the brotli tests in `zlibEmptyBuffer`

File: src/workerd/api/node/zlib-util.h
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:95]:
```suggestion
  explicit ZlibContext(ZlibMode _mode): mode(_mode) {}
```

File: src/workerd/api/node/zlib-util.h
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:182]:
```suggestion
  explicit BrotliContext(ZlibMode _mode): mode(_mode) {}
```

File: src/workerd/api/node/zlib-util.c++
Comment (npaun) [inline src/workerd/api/node/zlib-util.c++:443]:
```suggestion
template <typename CompressionContext>
```

There's no difference between `typename` and `class` in this context but we should be consistent throughout.

File: src/workerd/api/node/zlib-util.c++
Comment (npaun) [inline src/workerd/api/node/zlib-util.c++:449]:
```suggestion
template <typename CompressionContext>
```

**Comment by anonrig** [comment]
> GH won't let me comment on the line, but turn on the brotli tests in `zlibEmptyBuffer`

@npaun Those are the one-shot methods that this pull-request doesn't cover.

File: src/workerd/api/node/zlib-util.h
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:300]: Is this a pointer because node? I'd rather use a reference

**Comment by npaun** [comment]
@anonrig  Right, that's my job then hehe

File: src/workerd/api/node/zlib-util.h
Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:296]: This follows Node.js implementation but this is actually needed in `BrotliCompressionStream<CompressionContext>::initialize` function.

The use case is `this->context()->initialize` where if we define this context as a protected variable and expose it, the compiler will complain, since both BrotliCompressionStream and CompressionStream has context variables. (PS: BrotliCompressionStream extends CompressionStream)

File: src/workerd/api/node/zlib-util.h
Comment (npaun) [inline src/workerd/api/node/zlib-util.h:377]: Not sure what this method does because it should be inherited.

Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:377]: it returns the context of the inherited class. This won't work because the usage of excessive templates makes the compilers work extremely hard when used with `this->context.myFunction()`

File: src/workerd/api/node/zlib-util.h
Comment (npaun) [reply src/workerd/api/node/zlib-util.h:296]: I don't quite get why there would be two of them. I see only one variable named context_

Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:296]: I'll open a PR after this pull-request lands, to show the build error. For now, if this is not blocking, let's proceed with the PR. wdyt @npaun?

Comment (npaun) [reply src/workerd/api/node/zlib-util.h:296]: Not blocking. I'll start tagging my comments with _nit:_ for clarity.

**Comment by npaun** [comment]
Just flagging that there's a memory issue in the linux-asan build

File: src/workerd/api/node/zlib-util.h
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:255]:
```suggestion
    ~CompressionStream() noexcept(false);
```

Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:255]: Not possible:

```
src/workerd/api/node/zlib-util.h:260:5: error: exception specification is not available until end of class definition
  260 |     ~CompressionStream() noexcept(false);
```

**Review by npaun — APPROVED** [review]
LGTM but I'd wait for James || Felix to approve as well before merging because it's a bigger change

File: src/workerd/api/node/zlib-util.h
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:203]: Probably a good idea to add some extra comments at the top of this file pointing out that a lot of this was adapted from Node.js and therefore follows a lot of Node.js style as opposed to kj style. Some folks coming into this code later on might not have any idea who or what an "addaleax" is ;-)

Comment (anonrig) [reply src/workerd/api/node/zlib-util.h:203]: Agreed. Just added.

File: src/workerd/api/node/zlib-util.c++
Comment (jasnell) [inline src/workerd/api/node/zlib-util.c++:454]: Why did you take the `noexcept(false)` off this signature? It should be retained.

File: src/workerd/api/node/zlib-util.c++
Comment (jasnell) [inline src/workerd/api/node/zlib-util.c++:488]: Nit: Can `context()` return the reference instead of a pointer? ...`context().setBuffer(...)` just looks cleaner

**Review by jasnell — APPROVED** [review]

File: src/workerd/api/node/zlib-util.c++
Comment (anonrig) [reply src/workerd/api/node/zlib-util.c++:454]: It's producing the following error that requires some hacky solution to get around it:

```
src/workerd/api/node/zlib-util.h:260:5: error: exception specification is not available until end of class definition
  260 |     ~CompressionStream() noexcept(false);
```

File: src/workerd/api/node/zlib-util.c++
Comment (anonrig) [reply src/workerd/api/node/zlib-util.c++:488]: I'll follow up with a new pull-request for this.

File: src/workerd/api/node/zlib-util.h
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:305]: Our typical style would be to return a `CompressionContext&` here instead... and if you do want/need to provide the pointer, provide a separate `CompressionContext* get()` or something.

File: src/workerd/api/node/zlib-util.h
Comment (jasnell) [inline src/workerd/api/node/zlib-util.h:321]: non-blocking nit: I would expand the comment slight to connect this more concretely with the `AllocFor_` and `FreeFor_` APIs above.

**Pull request merged**
